src/core/analyzer-cache-context.ts

```diff
@@ -258,37 +252,74 @@ export class AnalyzerCacheContext {
     return copy;
   }
 
+  private async _scanLocal(resolvedUrl: string, contents?: string):
+      Promise<ScannedDocument> {
+    return this._cache.scannedDocumentPromises.getOrCompute(
+        resolvedUrl, async() => {
+          try {
+            const parsedDoc = await this._parse(resolvedUrl, contents);
+            const scannedDocument = await this._scanDocument(parsedDoc);
+
+            // Find all non-lazy imports
+            const imports = scannedDocument.getNestedFeatures().filter(
+                (e) => e instanceof ScannedImport &&
+                    e.type !== 'lazy-html-import') as ScannedImport[];
+
+            // Update dependency graph
+            const importUrls = imports.map((i) => this._resolveUrl(i.url));
+            this._cache.dependencyGraph.addDocument(resolvedUrl, importUrls);
+
+            return scannedDocument;
+          } catch (e) {
+            this._cache.dependencyGraph.rejectDocument(resolvedUrl, e);
+            throw e;
+          }
+        });
+  }
+
   /**
    * Scan a toplevel document and all of its transitive dependencies.
    */
-  private async _scan(
-      resolvedUrl: string, contents?: string,
-      visited?: Set<string>): Promise<ScannedDocument|'visited'> {
-    if (visited && visited.has(resolvedUrl)) {
-      return 'visited';
-    }
-
-    const actualVisited = visited || new Set();
-    actualVisited.add(resolvedUrl);
-    const scannedDocument =
-        await this._cache.scannedDocumentPromises.getOrCompute(
-            resolvedUrl, async() => {
-              const parsedDoc = await this._parse(resolvedUrl, contents);
-              return this._scanDocument(parsedDoc, actualVisited);
+  private async _scan(resolvedUrl: string, contents?: string):
+      Promise<ScannedDocument> {
+    return this._cache.dependenciesScannedPromises.getOrCompute(
+        resolvedUrl, async() => {
+          const scannedDocument = await this._scanLocal(resolvedUrl, contents);
+          // Find all non-lazy imports
+          const imports = scannedDocument.getNestedFeatures().filter(
+              (e) => e instanceof ScannedImport &&
+                  e.type !== 'lazy-html-import') as ScannedImport[];
+
+          // Scan imports
+          for (const scannedImport of imports) {
+            const importUrl = this._resolveUrl(scannedImport.url);
+            this._scan(importUrl).catch((error) => {
```

Comment:
Worth adding a comment here that we're intentionally not awaiting on the result here to avoid the deadlock, and we're getting the benefit of that await with the call to `dependencyGraph.whenReady` below.